File: 4_linked_list/2_middle_of_linked_list.py
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        if fast.next is None:
            return slow
        if fast.next.next is None:
            return slow.next

File: 04_linked_list/2_middle_of_linked_list.rs
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut slow, mut fast) = (&head, &head);
        loop {
            if let Some(node) = fast {
                fast = &node.next
            } else {
                return slow.clone();
            }
            if let Some(node) = fast {
                fast = &node.next
            } else {
                return slow.clone();
            }
            if let Some(node) = slow {
                slow = &node.next
            }
        }
    }
}
